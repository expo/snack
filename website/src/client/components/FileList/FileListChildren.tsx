import { StyleSheet, css } from 'aphrodite';
import escapeRegexp from 'escape-string-regexp';
import * as React from 'react';

import { SDKVersion } from '../../types';
import { isInsideFolder } from '../../utils/fileUtilities';
import { ThemeName } from '../Preferences/withThemeName';
import FileListEntry from './FileListEntry';
import { FileSystemEntry } from './types';

type Props = {
  parent: string;
  entries: FileSystemEntry[];
  clipboard: FileSystemEntry[];
  onOpen: (path: string) => void;
  onFocus: (path: string) => void;
  onSelect: (path: string) => void;
  onCopy: (path: string) => void;
  onDelete: (path: string) => void;
  onExpand: (path: string, expand?: boolean) => void;
  onCreateFile: (path: string | undefined) => void;
  onCreateFolder: (path: string | undefined) => void;
  onRename: (oldPath: string, newPath: string) => void;
  onPaste: (path: string | undefined, entry: FileSystemEntry) => void;
  onClearClipboard: () => void;
  sdkVersion: SDKVersion;
  theme: ThemeName;
  className?: string;
};

export default class FileListChildren extends React.PureComponent<Props> {
  _getImmediateChildren = () =>
    this.props.entries.filter(
      (e) =>
        // Filter-out non-immediate children
        !e.item.path.replace(new RegExp(`^${escapeRegexp(this.props.parent)}/`), '').includes('/')
    );

  render() {
    const {
      entries,
      clipboard,
      onCreateFile,
      onCreateFolder,
      onFocus,
      onOpen,
      onSelect,
      onPaste,
      onRename,
      onExpand,
      onDelete,
      onCopy,
      onClearClipboard,
      className,
      theme,
    } = this.props;

    return (
      <div className={`${css(styles.container)} ${className ?? ''}`}>
        {this._getImmediateChildren()
          .sort((a, b) => {
            if (a.item.type === b.item.type) {
              if (a.item.path.startsWith('.') && !b.item.path.startsWith('.')) {
                return 1;
              }

              if (b.item.path.startsWith('.') && !a.item.path.startsWith('.')) {
                return -1;
              }

              return a.item.path.localeCompare(b.item.path);
            } else if (a.item.type === 'folder') {
              return -1;
            } else {
              return 1;
            }
          })
          .map((e) => (
            <FileListEntry
              key={e.item.path}
              entry={e}
              rest={entries.filter((en) => isInsideFolder(en.item.path, e.item.path))}
              clipboard={clipboard}
              onCreateFile={onCreateFile}
              onCreateFolder={onCreateFolder}
              onOpen={onOpen}
              onSelect={onSelect}
              onFocus={onFocus}
              onCopy={onCopy}
              onPaste={onPaste}
              onDelete={onDelete}
              onRename={onRename}
              onExpand={onExpand}
              onClearClipboard={onClearClipboard}
              getAdjacentEntries={this._getImmediateChildren}
              sdkVersion={this.props.sdkVersion}
              theme={theme}
            />
          ))}
      </div>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    display: 'inline-block',
    width: '100%',
  },
});
