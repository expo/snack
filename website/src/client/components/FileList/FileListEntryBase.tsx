import { StyleSheet, css } from 'aphrodite';
import * as React from 'react';

import ContextMenu, { Action } from '../shared/ContextMenu';
import { isKeyCombo } from '../shared/KeybindingsManager';
import FileListEntryDropTarget from './FileListEntryDropTarget';
import { FileSystemEntry } from './types';

type Props = {
  entry: FileSystemEntry;
  rest: FileSystemEntry[];
  onOpen: (path: string) => void;
  onFocus: (path: string) => void;
  onRename: (oldPath: string, newPath: string) => void;
  onExpand?: (path: string, expand?: boolean) => void;
  renderMenuIcon: () => React.ReactNode;
  renderItem: () => React.ReactNode;
  renderTree?: () => React.ReactNode;
  actions: (Action | undefined)[];
  draggable?: boolean;
  theme: string | undefined;
};

type State = {
  menu: {
    pageX: number;
    pageY: number;
  } | null;
  isHovered: boolean;
};

export let lastDraggedEntry: FileSystemEntry | null;

export default class FileListEntry extends React.Component<Props, State> {
  static defaultProps = {
    draggable: true,
  };

  state = {
    menu: null,
    isHovered: false,
  };

  componentDidMount() {
    document.addEventListener('click', this._handleDocumentClick);
    document.addEventListener('contextmenu', this._handleDocumentContextMenu);
  }

  componentWillUnmount() {
    document.removeEventListener('click', this._handleDocumentClick);
    document.removeEventListener('contextmenu', this._handleDocumentContextMenu);
  }

  _handleClick = (event: React.MouseEvent<HTMLDivElement>) => {
    if ((event.target as HTMLDivElement).tagName === 'INPUT') {
      return;
    }

    this.props.onOpen(this.props.entry.item.path);
  };

  _handleMouseEnter = () =>
    this.setState({
      isHovered: true,
    });

  _handleMouseLeave = () =>
    this.setState({
      isHovered: false,
    });

  _handleMouseDown = () => (this._click = true);

  _handleFocus = () => {
    if (this._click) {
      // The focus was triggered by a click event
      // Ignore it to avoid double handling
      this._click = false;
      return;
    }

    this.props.onFocus(this.props.entry.item.path);
  };

  _handleKeyDown = (event: React.KeyboardEvent<HTMLDivElement>) => {
    if ((event.target as HTMLDivElement).tagName === 'INPUT') {
      return;
    }

    const bindings = this.props.actions.filter((action) =>
      action?.combo ? isKeyCombo(event as any, action.combo) : false
    );

    if (bindings.length) {
      event.preventDefault();

      bindings.forEach((binding) => binding?.handler());
    }
  };

  _hideContextMenu = () => this.setState({ menu: null });

  _showContextMenu = (e: MouseEvent) => {
    this.setState({
      menu: {
        pageX: e.pageX,
        pageY: e.pageY,
      },
    });
  };

  _handleDocumentClick = (e: MouseEvent) => {
    if (this.state.menu) {
      if (
        this._menu.current &&
        e.target !== this._menu.current &&
        !this._menu.current.contains(e.target as HTMLElement)
      ) {
        this._hideContextMenu();
      }
    } else if (
      this._more.current &&
      (e.target === this._more.current || this._more.current.contains(e.target as Node))
    ) {
      if (this.state.menu) {
        this._hideContextMenu();
      } else {
        this._showContextMenu(e);
      }
    }
  };

  _handleDocumentContextMenu = (e: MouseEvent) => {
    if (e.target === this._item.current || this._item.current?.contains(e.target as Node)) {
      e.preventDefault();
      this._showContextMenu(e);
    } else if (this.state.menu) {
      this._hideContextMenu();
    }
  };

  _handleDragStart = (e: React.DragEvent<HTMLDivElement>) => {
    e.dataTransfer.effectAllowed = 'move';
    e.dataTransfer.setData('application/json', JSON.stringify(this.props.entry));
    lastDraggedEntry = this.props.entry;
  };

  _handleDragEnd = () => {
    lastDraggedEntry = null;
  };

  _click: boolean = false;
  _item = React.createRef<HTMLDivElement>();
  _more = React.createRef<HTMLButtonElement>();
  _menu = React.createRef<HTMLUListElement>();

  render() {
    const { entry, rest, draggable, onRename, onExpand, theme, actions } = this.props;
    const { menu, isHovered } = this.state;

    return (
      <FileListEntryDropTarget entry={entry} rest={rest} onRename={onRename} onExpand={onExpand}>
        <div
          ref={this._item}
          draggable={draggable}
          onDragStart={this._handleDragStart}
          onDragEnd={this._handleDragEnd}
          tabIndex={0}
          className={css(styles.entry)}
          onClick={this._handleClick}
          onMouseEnter={this._handleMouseEnter}
          onMouseLeave={this._handleMouseLeave}
          onMouseDown={this._handleMouseDown}
          onFocus={this._handleFocus}
          onKeyDown={this._handleKeyDown}>
          {this.props.renderItem()}
        </div>
        <div
          className={css(
            styles.highlight,
            theme === 'dark' ? styles.highlightDark : styles.highlightLight,
            entry.state.isSelected ? styles.highlightActive : null
          )}
        />
        <ContextMenu
          ref={this._menu}
          visible={Boolean(menu)}
          position={menu}
          actions={actions}
          onHide={this._hideContextMenu}
        />
        <button
          ref={this._more}
          tabIndex={-1}
          className={css(
            styles.more,
            isHovered || menu ? styles.moreVisible : styles.moreInvisible
          )}>
          {this.props.renderMenuIcon()}
        </button>
        {this.props.renderTree?.()}
      </FileListEntryDropTarget>
    );
  }
}

const styles = StyleSheet.create({
  entry: {
    position: 'relative',
    display: 'inline-block',
    outline: 0,
    padding: '4px 16px',
    width: '100%',
    cursor: 'pointer',
    zIndex: 1,
    whiteSpace: 'nowrap',
  },

  highlight: {
    content: '""',
    display: 'inline-block',
    height: 30,
    position: 'absolute',
    left: 0,
    right: 0,
    opacity: 0,
    zIndex: 1,
    pointerEvents: 'none',
  },

  highlightLight: {
    backgroundColor: 'rgba(0, 0, 0, .04)',
  },

  highlightDark: {
    backgroundColor: 'rgba(255, 255, 255, .04)',
  },

  highlightActive: {
    opacity: 1,
  },

  more: {
    position: 'absolute',
    left: -8,
    border: 0,
    outline: 0,
    background: 'none',
    height: 30,
    zIndex: 2,
    padding: '7px 12px',
    textAlign: 'right',

    ':hover': {
      opacity: 1,
    },
  },

  moreInvisible: {
    opacity: 0,
  },

  moreVisible: {
    opacity: 1,
  },

  icon: {
    fill: 'currentColor',
    height: 16,
    width: 16,
  },
});
