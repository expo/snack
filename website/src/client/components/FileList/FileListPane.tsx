import { StyleSheet, css } from 'aphrodite';
import * as React from 'react';

type Props = {
  title: string;
  expanded: boolean;
  onClick: (event: React.MouseEvent<HTMLDivElement>) => void;
  buttons?: React.ReactNode[];
  children?: React.ReactNode;
  className?: string;
};

export default function FileListPane({
  title,
  expanded,
  buttons,
  children,
  onClick,
  className,
}: Props) {
  return (
    <div className={`${className ?? ''} ${css(styles.container)}`}>
      <div className={css(styles.header)}>
        <div
          className={css(styles.left)}
          onClick={onClick}
          data-test-id={`file-list-pane-${title.toLowerCase().replace(/[^a-z]/g, '-')}`}>
          <svg
            className={css(
              styles.collapse,
              expanded ? styles.collapseExpanded : styles.collapseCollapsed
            )}
            width="12px"
            height="12px"
            viewBox="0 0 12 12">
            <g
              stroke="none"
              strokeWidth="2"
              fill="none"
              fillRule="evenodd"
              strokeLinecap="round"
              strokeLinejoin="round">
              <polyline
                stroke="currentColor"
                transform="translate(6.000000, 6.000000) scale(-1, 1) rotate(180.000000) translate(-6.000000, -6.000000) "
                points="2 4 6 8 10 4"
              />
            </g>
          </svg>
          <h4 className={css(styles.label)}>{title}</h4>
        </div>
        <div className={css(styles.right)}>{buttons}</div>
      </div>
      {expanded ? children : null}
    </div>
  );
}

const styles = StyleSheet.create({
  container: {
    display: 'flex',
    flexDirection: 'column',
  },
  header: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    padding: '0 4px',
  },
  label: {
    fontSize: '1em',
    fontWeight: 500,
    lineHeight: 1,
    margin: 0,
  },
  collapse: {
    margin: '1px 6px 1px 2px',
    opacity: 0.7,
  },
  collapseCollapsed: {
    transform: 'rotate(180deg)',
  },
  collapseExpanded: {
    transform: 'rotate(0deg)',
  },
  left: {
    flex: 1,
    display: 'flex',
    flexDirection: 'row',
    padding: '8px 4px',
    cursor: 'pointer',
  },
  right: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
  },
});
