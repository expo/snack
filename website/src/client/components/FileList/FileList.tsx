import { StyleSheet, css } from 'aphrodite';
import pickBy from 'lodash/pickBy';
import * as React from 'react';
import { isModulePreloaded } from 'snack-sdk';

import eslintrc from '../../configs/eslint.json';
import { SDKVersion, SaveStatus, SnackFiles, SnackFile, Annotation } from '../../types';
import {
  isPackageJson,
  getUniquePath,
  isInsideFolder,
  isESLintConfig,
} from '../../utils/fileUtilities';
import type { EditorModal } from '../EditorViewProps';
import ImportExportMenu from '../Import/ImportExportMenu';
import ImportFilesManager from '../Import/ImportFilesManager';
import withThemeName, { ThemeName } from '../Preferences/withThemeName';
import SidebarShell from '../Shell/SidebarShell';
import { c } from '../ThemeProvider';
import ResizablePane from '../shared/ResizablePane';
import Toast from '../shared/Toast';
import FileListChildren from './FileListChildren';
import FileListEntryDropTarget from './FileListEntryDropTarget';
import FileListOpenEntry from './FileListOpenEntry';
import FileListPane from './FileListPane';
import FileListPaneButton from './FileListPaneButton';
import closeEntry from './actions/closeEntry';
import createNewEntry from './actions/createNewEntry';
import expandEntry from './actions/expandEntry';
import openEntry from './actions/openEntry';
import pasteEntry from './actions/pasteEntry';
import recursivelyCreateParents from './actions/recursivelyCreateParents';
import renameEntry from './actions/renameEntry';
import selectEntry from './actions/selectEntry';
import updateEntry from './actions/updateEntry';
import { FileSystemEntry, TextFileEntry, AssetFileEntry } from './types';
import { filesToEntries, findFocusedEntry } from './utils/convertFileStructure';

type Props = {
  visible: boolean;
  files: SnackFiles;
  selectedFile: string;
  updateFiles: (updateFn: (files: SnackFiles) => { [path: string]: SnackFile | null }) => void;
  onSelectFile: (path: string) => void;
  onRemoveFile: (path: string) => void;
  onRenameFile: (oldPath: string, newPath: string) => void;
  uploadFileAsync: (file: File) => Promise<string>;
  onDownloadCode: () => Promise<void>;
  onShowModal: (modal: EditorModal) => void;
  hasSnackId: boolean;
  saveStatus: SaveStatus;
  sdkVersion: SDKVersion;
  theme: ThemeName;
  annotations: Annotation[];
};

type State = {
  clipboard: FileSystemEntry[];
  deleted: {
    id: number;
    path: string;
    entries: FileSystemEntry[];
  }[];
  openFilesPane: boolean;
  projectPane: boolean;
  entries: FileSystemEntry[];
  selectedFile: string;
};

const AddIcon = () => (
  <g transform="translate(7.000000, 7.000000)">
    <circle fill={c('success')} cx="4.5" cy="4.5" r="4.5" />
    <rect fill={c('success-text')} x="4" y="2" width="1" height="5" />
    <rect fill={c('success-text')} x="2" y="4" width="5" height="1" />
  </g>
);

class FileList extends React.PureComponent<Props, State> {
  state: State = {
    clipboard: [],
    deleted: [],
    openFilesPane: true,
    projectPane: true,
    entries: [],
    selectedFile: '',
  };

  static getDerivedStateFromProps(props: Props, state: State) {
    let entries = filesToEntries(props.files, props.annotations, state.entries);
    entries =
      state.selectedFile !== props.selectedFile ? openEntry(entries, props.selectedFile) : entries;
    return {
      entries,
      selectedFile: props.selectedFile,
    };
  }

  private updateEntries(entries: FileSystemEntry[]) {
    const prevEntries = this.state.entries;
    this.setState({ entries });

    // Sync changes
    this.props.updateFiles((files) => {
      const updates: { [path: string]: SnackFile | null } = {};

      // Handle file removal (and rename)
      for (const path in files) {
        const entry = entries.find((entry) => entry.item.path === path);
        if (!entry) {
          updates[path] = null;
        }
      }

      // Handle added/renamed files
      entries.forEach((entry) => {
        if (
          entry.item.type === 'file' &&
          !files[entry.item.path] &&
          !isPackageJson(entry.item.path)
        ) {
          updates[entry.item.path] = {
            type: entry.item.asset ? 'ASSET' : 'CODE',
            contents: entry.item.asset ? entry.item.uri : entry.item.content,
          };
        }
      });

      return updates;
    });

    // Update focus
    const prevFocusedEntry = findFocusedEntry(prevEntries);
    const focusedEntry = findFocusedEntry(entries);
    if (focusedEntry?.item.path !== prevFocusedEntry?.item.path) {
      this.props.onSelectFile(focusedEntry?.item.path ?? '');
    }
  }

  _handleEntrySelect = (path: string) => {
    this.updateEntries(selectEntry(this.state.entries, path));
  };

  _handleEntryOpen = (path: string) => {
    this.updateEntries(openEntry(this.state.entries, path));
  };

  _handleEntryFocus = (path: string) => {
    this.updateEntries(openEntry(this.state.entries, path, true));
  };

  _handleEntryExpand = (path: string, expand?: boolean) =>
    this.updateEntries(expandEntry(this.state.entries, path, expand));

  _handleEntryRename = (oldPath: string, newPath: string) => {
    if (oldPath === newPath) {
      return;
    }

    const entry = this.state.entries.find((e) => e.item.path === oldPath);
    if (entry) {
      if (entry.item.type === 'folder') {
        this.state.entries
          .filter((e) => e.item.type === 'file' && e.item.path.startsWith(oldPath))
          .forEach((renamedFile) => {
            this.props.onRenameFile(
              renamedFile.item.path,
              `${newPath}${renamedFile.item.path.substring(oldPath.length)}`
            );
          });
      } else {
        this.props.onRenameFile(oldPath, newPath);
      }
    }

    this.updateEntries(renameEntry(this.state.entries, oldPath, newPath));
  };

  _restoreEntries = (entries: FileSystemEntry[]) =>
    this.updateEntries([
      ...this.state.entries,
      ...entries.map((e) =>
        updateEntry(e, {
          item: {
            path: getUniquePath(
              this.state.entries.map((it) => it.item.path),
              e.item.path
            ),
          },
        })
      ),
    ]);

  _handleDismissDelete = (id: number) =>
    this.setState((state) => ({
      deleted: state.deleted.filter((g) => g.id !== id),
    }));

  _handleEntryClose = (path: string) =>
    this.updateEntries(this.state.entries.map((e) => (e.item.path === path ? closeEntry(e) : e)));

  _handleEntryCloseOthers = (path: string) =>
    this.updateEntries(this.state.entries.map((e) => (e.item.path !== path ? closeEntry(e) : e)));

  _handleEntryCloseAll = () => this.updateEntries(this.state.entries.map((e) => closeEntry(e)));

  _handleEntryDelete = (path: string) => {
    const entry = this.state.entries.find((e) => e.item.path === path);
    const deletedEntries: FileSystemEntry[] = [];

    this.updateEntries(
      this.state.entries.filter((e) => {
        const remove = e.item.path === path || isInsideFolder(e.item.path, path);
        if (remove) {
          deletedEntries.push(e);
          if (e.item.type === 'file') {
            this.props.onRemoveFile(e.item.path);
          }
        }
        return !remove;
      })
    );

    this.setState((state) => ({
      deleted: [
        ...state.deleted,
        {
          id: this._currentDeleteID++,
          path: entry ? entry.item.path : 'Item',
          entries: deletedEntries,
        },
      ],
    }));
  };

  _currentDeleteID: number = 0;

  _handleEntryImport = (entry: TextFileEntry | AssetFileEntry) => {
    let entries: FileSystemEntry[];

    if (isPackageJson(entry.item.path)) {
      // Merge dependencies from package.json file
      entries = this.state.entries.map((e) => {
        if (isPackageJson(e.item.path)) {
          try {
            // @ts-ignore
            const previous = JSON.parse(e.item.content);
            // @ts-ignore
            const next = JSON.parse(entry.item.content);

            return {
              ...e,
              item: {
                ...e.item,
                content: JSON.stringify(
                  {
                    ...previous,
                    dependencies: {
                      ...previous.dependencies,
                      ...Object.keys(next.dependencies).reduce(
                        (acc: { [key: string]: string }, name) => {
                          if (!isModulePreloaded(name, this.props.sdkVersion, true)) {
                            acc[name] = next.dependencies[name];
                          }

                          return acc;
                        },
                        {}
                      ),
                    },
                  },
                  null,
                  2
                ),
              },
            } as TextFileEntry;
          } catch (err) {
            // Do nothing
          }
        }

        return e;
      });
    } else if (isESLintConfig(entry.item.path)) {
      try {
        const content = JSON.parse((entry as TextFileEntry).item.content);

        // Cleanup the config file to remove unsupported plugins and rules
        if (content.plugins) {
          content.plugins = content.plugins.filter((name: string) =>
            eslintrc.plugins.includes(name)
          );

          if (!content.plugins.length) {
            delete content.plugins;
          }
        }

        if (content.rules) {
          content.rules = pickBy(content.rules, (_, key) => {
            if (key.includes('/')) {
              return eslintrc.plugins.some((name) => key.startsWith(`${name}/`));
            }

            return key;
          });
        }

        if (content.extends) {
          if (typeof content.extends === 'string' && !content.extends.startsWith('eslint:')) {
            delete content.extends;
          } else {
            content.extends = content.extends.filter((name: string) => name.startsWith('eslint:'));

            if (!content.extends.length) {
              delete content.extends;
            }
          }
        }

        // Remove existing eslintrc if any
        entries = this.state.entries.filter((e) => !isESLintConfig(e.item.path));
        entries.push(
          updateEntry(entry, {
            item: {
              content: JSON.stringify(content, null, 2),
            },
          })
        );
      } catch (e) {
        // Ignore errors
        entries = this.state.entries;
      }
    } else {
      const parents = recursivelyCreateParents(this.state.entries, entry.item.path);

      entries = [...this.state.entries, ...parents];
      entries.push(
        updateEntry(entry, {
          item: {
            path: getUniquePath(
              entries.map((e) => e.item.path),
              entry.item.path
            ),
          },
        })
      );
    }

    this.updateEntries(entries);
  };

  _handleEntryPaste = (path: string | undefined, e: FileSystemEntry) =>
    this.updateEntries(pasteEntry(this.state.entries, path, e));

  _handleCopy = (path: string) =>
    this.setState((state) => ({
      clipboard: state.entries.filter((e) => e.item.path === path),
    }));

  _handleClearClipboard = () =>
    this.setState({
      clipboard: [],
    });

  _toggleOpenFilesPane = () => this.setState((state) => ({ openFilesPane: !state.openFilesPane }));

  _toggleProjectPane = () => this.setState((state) => ({ projectPane: !state.projectPane }));

  _handleCreateFile = (path?: string | undefined) => {
    let entries = createNewEntry(this.state.entries, 'file', path);
    const newEntry = entries.find(
      (e) => !this.state.entries.find((e2) => e2.item.path === e.item.path)
    );
    entries = newEntry ? openEntry(entries, newEntry.item.path, true) : entries;
    this.updateEntries(entries);
  };

  _handleCreateFolder = (path?: string | undefined) => {
    this.updateEntries(createNewEntry(this.state.entries, 'folder', path));
  };

  render() {
    return (
      <ImportFilesManager
        className={css(styles.container)}
        entries={this.state.entries}
        onImportFile={this._handleEntryImport}
        uploadFileAsync={this.props.uploadFileAsync}
        render={({ onImportStart }) =>
          this.props.visible ? (
            <ResizablePane direction="horizontal" className={css(styles.pane)}>
              <SidebarShell>
                <FileListPane
                  className={css(
                    this.state.projectPane ? styles.openFilesSmall : styles.openFilesLarge
                  )}
                  title="Open files"
                  expanded={this.state.openFilesPane}
                  onClick={this._toggleOpenFilesPane}>
                  <ul className={css(styles.tabs)} data-test-id="file-list-open-files-content">
                    {this.state.entries
                      .filter((e) => e.item.type === 'file' && e.state.isOpen)
                      .map((e: any) => (
                        <FileListOpenEntry
                          key={e.item.path}
                          entry={e}
                          onOpen={() => this._handleEntryOpen(e.item.path)}
                          onClose={() => this._handleEntryClose(e.item.path)}
                          onCloseOthers={() => this._handleEntryCloseOthers(e.item.path)}
                          onCloseAll={this._handleEntryCloseAll}
                        />
                      ))}
                  </ul>
                </FileListPane>
                <FileListPane
                  className={css(styles.project)}
                  title="Project"
                  expanded={this.state.projectPane}
                  onClick={this._toggleProjectPane}
                  buttons={[
                    <FileListPaneButton key="create-file" onClick={() => this._handleCreateFile()}>
                      <path
                        fillOpacity="0.7"
                        d="M3,2 L13,2 L13,14 L3,14 L3,2 Z M9,2 L13,6 L13,2 L9,2 Z M9,6 L9,2 L8,2 L8,7 L13,7 L13,6 L9,6 Z"
                      />
                      <AddIcon />
                    </FileListPaneButton>,
                    <FileListPaneButton
                      key="create-folder"
                      onClick={() => this._handleCreateFolder()}>
                      <path
                        fillOpacity="0.7"
                        d="M7.25,4 L7.5,4 L7.5,3 L7,3.5 L7,2 L15,2 L15,4 L7.25,4 Z M6.75,4 L5,4 L7,2 L7,3.5 L6.5,4 L6.75,4 Z M1,4 L15,4 L15,14 L1,14 L1,4 Z M7.5,3 L7.5,4 L14,4 L14,3 L7.5,3 Z"
                      />
                      <AddIcon />
                    </FileListPaneButton>,
                    <ImportExportMenu
                      key="menu"
                      onImportFilesClick={onImportStart}
                      onImportRepoClick={() => this.props.onShowModal('import-repo')}
                      onImportProductionClick={() => this.props.onShowModal('import-production')}
                      onExportClick={this.props.onDownloadCode}
                      saveStatus={this.props.saveStatus}
                      hasSnackId={this.props.hasSnackId}
                    />,
                  ]}>
                  <FileListEntryDropTarget
                    className={css(styles.files)}
                    rest={this.state.entries}
                    onRename={this._handleEntryRename}>
                    <div className={css(styles.children)} data-test-id="file-list-project-content">
                      <FileListChildren
                        parent=""
                        entries={this.state.entries}
                        clipboard={this.state.clipboard}
                        onCreateFile={this._handleCreateFile}
                        onCreateFolder={this._handleCreateFolder}
                        onOpen={this._handleEntryOpen}
                        onSelect={this._handleEntrySelect}
                        onFocus={this._handleEntryFocus}
                        onPaste={this._handleEntryPaste}
                        onRename={this._handleEntryRename}
                        onExpand={this._handleEntryExpand}
                        onDelete={this._handleEntryDelete}
                        onCopy={this._handleCopy}
                        onClearClipboard={this._handleClearClipboard}
                        sdkVersion={this.props.sdkVersion}
                        theme={this.props.theme}
                        className={css(styles.list)}
                      />
                    </div>
                  </FileListEntryDropTarget>
                </FileListPane>
                {this.state.deleted
                  .map((group) => (
                    <Toast
                      key={group.id}
                      label={`Deleted ${group.path.split('/').pop()}`}
                      actions={[
                        {
                          label: 'Undo',
                          action: () => {
                            this._restoreEntries(group.entries);
                            this._handleDismissDelete(group.id);
                          },
                        },
                        { label: 'Dismiss' },
                      ]}
                      onDismiss={() => this._handleDismissDelete(group.id)}
                    />
                  ))
                  .reverse()}
              </SidebarShell>
            </ResizablePane>
          ) : null
        }
      />
    );
  }
}

export default withThemeName(FileList);

const styles = StyleSheet.create({
  container: {
    display: 'flex',
    flexDirection: 'column',
  },
  list: {
    padding: '0 12px',
    height: '100%',
  },
  pane: {
    flex: 1,
    display: 'flex',
    flexDirection: 'column',
    width: 240,
    minWidth: 240,
    height: '100%',
    zIndex: 10,
  },
  project: {
    flex: 1,
  },
  openFilesSmall: {
    maxHeight: '50%',
  },
  openFilesLarge: {
    maxHeight: 'calc(100% - 30px)',
  },
  files: {
    flex: '1 0 0',
    overflow: 'auto',
    paddingBottom: 16,
  },
  children: {
    position: 'relative',
  },
  tabs: {
    margin: 0,
    listStyle: 'none',
    padding: '8px 0',
    overflow: 'auto',

    ':empty': {
      display: 'none',
    },
  },
  toolbar: {
    padding: 8,
  },
  toasts: {
    position: 'fixed',
    bottom: '3em',
    left: '1em',
    zIndex: 10,
  },
});
