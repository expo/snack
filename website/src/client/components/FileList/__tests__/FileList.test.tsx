import { StyleSheetTestUtils } from 'aphrodite';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import * as React from 'react';

import { TEST_SDK_VERSION } from '../../../configs/sdk';
import FileList from '../FileList';

// @ts-ignore
jest.mock('../../Preferences/withThemeName', () => (c) => c);

beforeEach(StyleSheetTestUtils.suppressStyleInjection);
afterEach(StyleSheetTestUtils.clearBufferAndResumeStyleInjection);

it('renders children', () => {
  const wrapper = shallow(
    <FileList
      files={{
        'App.js': {
          type: 'CODE',
          contents: '',
        },
        'components/Icon.js': {
          type: 'CODE',
          contents: '',
        },
      }}
      selectedFile="App.js"
      updateFiles={jest.fn()}
      onRemoveFile={jest.fn()}
      onRenameFile={jest.fn()}
      onSelectFile={jest.fn()}
      uploadFileAsync={jest.fn()}
      onDownloadCode={jest.fn()}
      onShowModal={jest.fn()}
      hasSnackId={false}
      saveStatus="edited"
      sdkVersion={TEST_SDK_VERSION}
      visible
      annotations={[]}
    />
  );

  expect(toJSON(wrapper)).toMatchSnapshot();
});

it('deletes entry', () => {
  const updateFiles = jest.fn();
  const wrapper = shallow(
    <FileList
      files={{
        'test/App.js': {
          type: 'CODE',
          contents: '',
        },
      }}
      selectedFile="test/App.js"
      updateFiles={updateFiles}
      onRemoveFile={jest.fn()}
      onRenameFile={jest.fn()}
      onSelectFile={jest.fn()}
      uploadFileAsync={jest.fn()}
      onDownloadCode={jest.fn()}
      onShowModal={jest.fn()}
      hasSnackId={false}
      saveStatus="edited"
      sdkVersion={TEST_SDK_VERSION}
      visible
      annotations={[]}
    />
  );

  // @ts-ignore
  wrapper.instance()._handleEntryDelete('test');

  expect(updateFiles).toBeCalledTimes(1);
});

it('copies item to clipboard', () => {
  const wrapper = shallow(
    <FileList
      files={{
        'test/App.js': {
          type: 'CODE',
          contents: '',
        },
      }}
      selectedFile="App.js"
      updateFiles={jest.fn()}
      onRemoveFile={jest.fn()}
      onRenameFile={jest.fn()}
      onSelectFile={jest.fn()}
      uploadFileAsync={jest.fn()}
      onDownloadCode={jest.fn()}
      onShowModal={jest.fn()}
      hasSnackId={false}
      saveStatus="edited"
      sdkVersion={TEST_SDK_VERSION}
      visible
      annotations={[]}
    />
  );

  // @ts-ignore
  expect(wrapper.state().clipboard).toEqual([]);

  // @ts-ignore
  wrapper.instance()._handleCopy('test/App.js');

  // @ts-ignore
  expect(wrapper.state().clipboard).toEqual([
    {
      item: { path: 'test/App.js', type: 'file', content: '', virtual: false },
      state: {
        isError: false,
        isFocused: false,
        isLoading: false,
        isOpen: false,
        isSelected: false,
      },
    },
  ]);

  // @ts-ignore
  wrapper.instance()._handleClearClipboard();

  // @ts-ignore
  expect(wrapper.state().clipboard).toEqual([]);
});
