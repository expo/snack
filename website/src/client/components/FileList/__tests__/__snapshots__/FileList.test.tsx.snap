// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders children 1`] = `
<ImportFilesManager
  className="container_wgmchy"
  entries={
    Array [
      Object {
        "item": Object {
          "content": "",
          "path": "App.js",
          "type": "file",
          "virtual": false,
        },
        "state": Object {
          "isError": false,
          "isFocused": true,
          "isLoading": false,
          "isOpen": true,
          "isSelected": true,
        },
      },
      Object {
        "item": Object {
          "path": "components",
          "type": "folder",
        },
        "state": Object {},
      },
      Object {
        "item": Object {
          "content": "",
          "path": "components/Icon.js",
          "type": "file",
          "virtual": false,
        },
        "state": Object {
          "isError": false,
          "isFocused": false,
          "isLoading": false,
          "isOpen": false,
          "isSelected": false,
        },
      },
    ]
  }
  onImportFile={[Function]}
  render={[Function]}
  uploadFileAsync={[MockFunction]}
/>
`;
